### PR TITLE
Define runner.py

### PR DESCRIPTION
**Pull request by Junefen**
The runner.py is used to define how to run virt-who by cli and service, and how to analyse the rhsm log.

Mainly include below three functions.
```
- run_virtwho_cli()
- run_virtwho_service()
- log_analyzer()
```

### CODE REVIEW

**Review by eko999 — COMMENTED** [review]
Two suggestions from me:
1). each class or function should have the doc string
2). think about and provide a demo test file, such as test_cli.py, test_config.py, how do we design the test cases ,and it will determine how to design the runner

File: virtwho/runner.py [hunk 1]
```diff
@@ -1,5 +1,387 @@
...
+        raise FailException("Failed to run virt-who service.")
+
+    def virtwho_thread_start(self, cli, wait):
+        q = queue.Queue()
+        self.rhsm_log_clean()
+        t1 = threading.Thread(target=self.virtwho_run, args=(q, cli))
+        t1.setDaemon(True)
+        t1.start()
+        rhsm_ouput = self.rhsm_log_get(wait)
+        return rhsm_ouput
```
Comment (Junefen) [inline virtwho/runner.py:116]: t1.setDaemon(True) can help to stop the sub-thread t1 when ```rhsm_ouput = self.rhsm_log_get(wait)``` and ```        return rhsm_ouput``` complete.

File: virtwho/runner.py [hunk 2]
```diff
@@ -1,5 +1,387 @@
...
+        data = dict()
+        data["debug"] = self.msg_search(rhsm_log, "\\[.*DEBUG\\]")
+        data["oneshot"] = self.msg_search(rhsm_log, "virt-who terminated")
+        data["send_number"] = self.virtwho_send_number(rhsm_log)
+        data["error_number"], data["error_list"] = self.virtwho_error_status()
+        data["report_id"] = self.virtwho_report_id(rhsm_log)
+        data["interval_time"] = self.virtwho_interval_time(rhsm_log)
+        data["loop_number"], data["loop_time"] = self.virtwho_loop_status()
+        data["mappings"] = self.virtwho_mappings(rhsm_log)
+        return data
```
Comment (Junefen) [inline virtwho/runner.py:85]: Will keep to add new content when find requirements

File: virtwho/runner.py [hunk 3]
```diff
@@ -46,19 +47,19 @@ def run_virtwho_cli(self,
             test interval function.
         :return: tty_output
         """
-        cmd = 'virt-who '
+        cmd = "virt-who "
```
Comment (eko999) [inline virtwho/runner.py:50]: 'xxx' is more better format, any issue for it?

File: virtwho/runner.py [hunk 4]
```diff
@@ -68,207 +69,111 @@ def run_virtwho_service(self, wait=None):
...
-            logger.info(f'The loop time get from log is {loop_time_get}')
-            analyzer_list.append(loop_time_get == loop_time)
-        if 'False' in analyzer_list:
-            return False
-        else:
-            return True
+    def log_analyzer(self, rhsm_log):
+        data = dict()
+        data["debug"] = self.msg_search(rhsm_log, "\\[.*DEBUG\\]")
+        data["oneshot"] = self.msg_search(rhsm_log, "virt-who terminated")
```
Comment (eko999) [inline virtwho/runner.py:78]: for the oneshot validation, I think only 'virt-who terminated' is not enough, sometimes, virt-who will be terminated by errors.

File: virtwho/runner.py [hunk 5]
```diff
@@ -68,207 +69,111 @@ def run_virtwho_service(self, wait=None):
...
-            analyzer_list.append(loop_number_get == loop_number)
-        if loop_time:
-            loop_time_get = self.vw_loop_time()
-            logger.info(f'The loop time get from log is {loop_time_get}')
-            analyzer_list.append(loop_time_get == loop_time)
-        if 'False' in analyzer_list:
-            return False
-        else:
-            return True
+    def log_analyzer(self, rhsm_log):
```
Comment (eko999) [inline virtwho/runner.py:75]: It would be very useful to provide explanation (doc string) for these keys in data dict.

**Review by eko999 — APPROVED** [review]
Thank you @Junefen , I like this new design very much, it's very clear for each function.

File: virtwho/runner.py (hunk 3 above)
Comment (Junefen) [reply virtwho/runner.py]: No issue, just want to make a uniform. And yes, it looks better to use '' than "" somewhere.

File: virtwho/runner.py (hunk 4 above)
Comment (Junefen) [reply virtwho/runner.py:83]: Yes, it's really true, now my thought is to add asserting ```result["send_number"] == 1 and result["error_number"] == 0``` when test oneshot function.
And besides above I think only check the key words ```virt-who terminated``` is also not enough, so added ```data["thread_number"] = self.virtwho_thread_number()``` to check ```result["thread_number"] == 0``` to make sure virt-who service is really stopped.

File: virtwho/runner.py [hunk 6]
```diff
@@ -1,5 +1,529 @@
...
+        thread_num = 0
+        cmd = "ps -ef | grep virt-who -i | grep -v grep |wc -l"
+        ret, output = self.ssh.runcmd(cmd)
+        if output is not None and output != "":
+            thread_num = int(output.strip())
+        return thread_num
+
+    def run_service(self, name='virt-who', action='restart'):
+        """
+
```
Comment (hkx303) [inline virtwho/runner.py:461]: seems this empty line is redundant or miss the description...

File: tests/function/test_cli.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        assert (result["send_number"] == 1
+                and result["interval_time"] == 3600)
+
+        # run with '-i 60' to test interval=60
+        result = runner.run_virtwho_cli(oneshot=False, interval=60, wait=60)
+        assert (result["send_number"] == 1
+                and result["interval_time"] == 60
+                and result["loop_time"] == 60)
+
+    def test_print_in_cli(self):
```
Comment (hkx303) [inline tests/function/test_cli.py:67]: Miss the docstring :)

**Review by hkx303 — APPROVED** [review]
ACK, thanks for PR, a great progress again.

File: tests/function/test_cli.py (hunk 7 above)
Comment (Junefen) [reply tests/function/test_cli.py]: Didn't give enough docstring for the demos, will perfect them when define the cases.

File: virtwho/runner.py (hunk 6 above)
Comment (Junefen) [reply virtwho/runner.py]: Done

File: virtwho/runner.py (hunk 5 above)
Comment (Junefen) [reply virtwho/runner.py]: Done

**Pull request merged**